src/BSON/Value.stub.php

```diff
@@ -0,0 +1,206 @@
+     * @var int
+     * @cvalue BSON_TYPE_BINARY
+     */
+    public const TYPE_BINARY = 0x05;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_UNDEFINED
+     */
+    public const TYPE_UNDEFINED = 0x06;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_OID
+     */
+    public const TYPE_OID = 0x07;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_BOOL
+     */
+    public const TYPE_BOOL = 0x08;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_DATE_TIME
+     */
+    public const TYPE_DATE_TIME = 0x09;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_NULL
+     */
+    public const TYPE_NULL = 0x0A;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_REGEX
+     */
+    public const TYPE_REGEX = 0x0B;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_DBPOINTER
+     */
+    public const TYPE_DBPOINTER = 0x0C;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_CODE
+     */
+    public const TYPE_CODE = 0x0D;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_SYMBOL
+     */
+    public const TYPE_SYMBOL = 0x0E;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_CODEWSCOPE
+     */
+    public const TYPE_CODEWSCOPE = 0x0F;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_INT32
+     */
+    public const TYPE_INT32 = 0x10;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_TIMESTAMP
+     */
+    public const TYPE_TIMESTAMP = 0x11;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_INT64
+     */
+    public const TYPE_INT64 = 0x12;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_DECIMAL128
+     */
+    public const TYPE_DECIMAL128 = 0x13;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_MAXKEY
+     */
+    public const TYPE_MAXKEY = 0x7F;
+
+    /**
+     * @var int
+     * @cvalue BSON_TYPE_MINKEY
+     */
+    public const TYPE_MINKEY = 0xFF;
+
+    private function __construct() {}
+
+    final public function getType(): int {}
+
+#if PHP_VERSION_ID >= 80000
+    final public function getValue(): mixed {}
+#else
+    /** @return mixed */
+    final public function getValue() {}
+#endif
+
+#if PHP_VERSION_ID >= 80000
+    final static public function fromPHP(mixed $value): Value {}
+#else
+    /** @param mixed $value */
+    final static public function fromPHP($value): Value {}
+#endif
+
+    /** Accessors */
+    final public function getArray(): PackedArray {}
+    final public function getBinary(): Binary {}
+    final public function getBool(): bool {}
+    final public function getCode(): Javascript {}
+    final public function getDBPointer(): DBPointer {}
+    final public function getDecimal128(): Decimal128 {}
+    final public function getDocument(): Document {}
+    final public function getDouble(): float {}
+    final public function getInt(): int {}
+    final public function getInt32(): int {}
+#if PHP_VERSION_ID >= 80000
+    final public function getInt64(): int|Int64 {}
+#else
+    /** @return int|Int64 */
+    final public function getInt64() {}
+#endif
```

Comment:
https://github.com/mongodb/mongo-php-driver/pull/1412#discussion_r1176193974 just made me think bank to this.

Is `getInt(): int` and `getInt32(): int` redundant? Or will they have different logic internally where `getInt()` warns on truncation and `getInt32()` does not?

I have more thoughts on `getInt64(): int|Int64`. If the method behaves just like `getInt()` except that it returns an Int64 instead of a truncation warning, then it's only relevant to 32-bit platforms. Such a method would always return `int` on 64-bit platforms and it'd effectively be an alias of `getInt()`.

Does it make more sense to have `getInt64()` always return Int64? There's a need on both 32-bit and 64-bit platforms to maintain type fidelity for 64-bit integers with values within the range of signed 32-bit integers. I do realize a user could just keep the Value as-is and pass it along for BSON encoding, so maybe this _isn't_ important.

Thoughts?

Reply:
All of the `get*` methods check the internal type and throw if that isn't the case. For example, calling `getInt64` on an instance storing an int32 will throw:
> Uncaught MongoDB\Driver\Exception\UnexpectedValueException: Unexpected BSON type 0x10

This is similar to `bson_iter_t`, which also refuses to coerce in its getters (e.g. `bson_iter_int64`), but includes specific methods to coerce types (e.g. `bson_iter_as_int64`). While the iterator class returns `0` (or nothing in case of `bson_iter_document`) when types do not match, we explicitly throw, which I believe is fine considering the API allows for checking types first.

This brings me to `getInt`, `getInt32`, and `getInt64`. Let's take the following code:
```php
$int32 = MongoDB\BSON\Value::fromPHP(2);
var_dump($int32->isInt());
var_dump($int32->isInt32());
var_dump($int32->isInt64());

var_dump($int32->getValue());
var_dump($int32->getInt());
var_dump($int32->getInt32());
var_dump($int32->getInt64());
```

The output of this is as follows:
```
bool(true)
bool(true)
bool(false)
int(2)
int(2)
int(2)
Uncaught MongoDB\Driver\Exception\UnexpectedValueException: Unexpected BSON type 0x10
```

To summarise, the consumer should use:
* `getValue` when the type does not matter at all and they are only interested in the value
* `get*` (e.g. `getInt32` or `getDocument`) when they want to ensure a specific BSON type; the corresponding `is` method should be used to avoid exceptions
* `getInt` when they want to use an integer but don't care whether it's 32-bit or 64-bit.

`getInt` is a special case here, but it behaves different from `getInt32` and `getInt64` as it only throws if the type is not int32 or int64, whereas the type-specific getters only check for the specific type they are designed for. I hope this clears up the confusion between `getInt` and `getInt32`.

---

As for `getInt64` always returning an `Int64` instance, this would definitely be an option. If we decide that `getInt64` should always return an object, I'd like to ensure that PHPC-1949 (opening the constructor of `Int64`) is implemented and that operations on an `Int64` work the same way they do with scalar integers. That would allow us to immediately define the return type of `getInt64` as `Int64` instead of the current `int | Int64`, avoiding an unnecessary change down the line.

The only question that remains then is how `getInt` should behave when returning the value of an int64 `bson_value_t` on 64-bit platforms: should it return an `int` for convenience or should it return an `Int64` for consistence.

Reply:
Would `getValue()` only return an Int64 iff we're on a 32-bit platform _and_ the value is beyond the 32-bit range? If it's meant to be a convenience function, I assume we'd want to return a scalar int whenever possible (e.g. `int64(0)` would not yield an Int64 object and folks can use `getInt64()` if they care about that).

> As for getInt64 always returning an Int64 instance, this would definitely be an option

I'm in favor of this, provided we can ensure Int64 objects can be cast to PHP integers. I don't believe that's tested/implemented now, but I created [PHPC-2213](https://jira.mongodb.org/browse/PHPC-2213) to track that.

> operations on an Int64 work the same way they do with scalar integers

Created [PHPC-2212](https://jira.mongodb.org/browse/PHPC-2212) to track that.

> how `getInt()` should behave

I'd need to give this more thought before we commit, but if `getValue()` is already a convenience method I'm not sure what the real use case for `getInt()` is.

 * If it only returns Int64 when absolutely necessary (i.e. values in 64-bit range on a 32-bit platform) then it's basically `getValue()` with additional type checks.
 * If it always returns a PHP integer and warns on truncation then it's probably not usable on 32-bit platforms (warnings could be unpredictable). On a 64-bit platform, it's still basically `getValue()` with type checking, expect you get the benefit of an `int` return type hint (is that terribly useful for static analysis?).
 * If it always returns BSON int32 as a PHP int and BSON int64 as an Int64 object, then it seems like a more pure alias of the `getInt32()` and `getInt64()` methods. Combined with casting support mentioned above, x64 users could do `(int) $value->getInt()`, but then this is basically no different than the preceding bullet.

How do you see this method being used?

Reply:
`isInt()` and `getInt()` have been removed. I didn't create a ticket yet as I agree with your point that it's just `getValue()` with added type checks, so there is no reason to have this method.

For integers, this leaves us with:
* `getInt32()`, which will always return an int value but throw if the value is an int64
* `getInt64()`, which will always return an `Int64` instance but throw if the value is an int32 (i.e., not `{$numberLong: ... }` in extended JSON). This is true on all platforms and for all values, so even int64 values that fall in the 32-bit range are returned as `Int64` instances.
* `getValue()`, which will return whatever the corresponding type specific getter would return.